### DISCUSSION TITLE
How to Compile DXWrapper Source

### DISCUSSION THREAD

**Discussion by MAXBURAOT** [elishacloud/dxwrapper#123] [Q&A] [answered]
-How to Compile DXWrapper Source, Currently im using Visual Studio 2019
-Is it ok to retarget the solution or stick with Visual Studio 2017 Tools

**Comment by elishacloud** [comment]
You are welcome to retarget the solution on your end and compile the code in Visual Studio 2019.  However, I want to keep the code in GitHub using Visual Studio 2017 for now.

The reason is because I have a number of projects I am working on and all of them currently use Visual Studio 2017.  I would like to continue to use the same version of Visual Studio for all these projects for as long as it makes sense.

However, if there is some technical reason why moving the solution to Visual Studio 2019 will help the project I am happy to reconsider.

**Reply by mirh** [reply]
I'd actually argue there is a technical reason to keep VS2017 as much as possible. 
Or.. I mean, I guess solution version is not exactly the same of toolset's, but keeping v141_xp around is certainly going to benefit some special use case.

**Reply by elishacloud** [reply]
Yeah, the v141_xp option is why I am still using VS2017.  I don't use that option on dxwrapper at this time but I do use that option on some of my other projects.

**Reply by MAXBURAOT** [reply]
Hi all, i tried to compile  the source, any idea about the errors, did i missed something like dependecies or etc :) Thanks for the replies 

![00](https://user-images.githubusercontent.com/33173384/115144736-ec8f3b80-a080-11eb-8791-970d8a3092f9.png)
![01](https://user-images.githubusercontent.com/33173384/115144739-edc06880-a080-11eb-977a-ee7ed1978622.png)
![02](https://user-images.githubusercontent.com/33173384/115145114-9ae7b080-a082-11eb-921c-74a59c12573d.png)

**Reply by elishacloud** [reply]
From the error messages it looks like you have not downloaded the submodules.  Check the `External` folder and update all the applicable submodules for each item in this folder.  Also see here: https://github.blog/2016-02-01-working-with-submodules/

**Reply by elishacloud** [reply]
You also need to download and install both the Windows 10 SDK and WDK. The exact version required can be seen in the project properties in Visual Studio.

The WDK includes the missing `d3dumddi.h` file:  https://docs.microsoft.com/en-us/windows-hardware/drivers/download-the-wdk

**Comment by elishacloud** [comment, marked as answer]
I was not planning to move to Visual Studio 2019, but I had to move to it to add the updated DDrawCompat v0.3.0 to the project.

To compile the latest build you need to have the latest SDK and WDK, both of which can be downloaded here:  https://docs.microsoft.com/en-us/windows-hardware/drivers/download-the-wdk

**Reply by MAXBURAOT** [reply]
Hi elisha, thank you for updating it, now no more errors on building it thanks 

![DX 01](https://user-images.githubusercontent.com/33173384/115745360-ef15cc00-a3c5-11eb-932b-73730a06bc9c.png)
![DX 02](https://user-images.githubusercontent.com/33173384/115745374-f2a95300-a3c5-11eb-8db0-aebfcdb05347.png)
